torch/quantization/fx/quantize.py

```diff
@@ -0,0 +1,765 @@
+import torch
+from torch.quantization import (
+    propagate_qconfig_,
+    convert,
+    DEFAULT_QAT_MODULE_MAPPING,
+)
+
+from torch.fx import (
+    GraphModule,
+)
+
+from torch.fx.graph import (
+    Graph,
+    Node,
+    map_arg,
+)
+
+from torch.fx.symbolic_trace import (
+    DefaultDelegate,
+)
+
+from .pattern_utils import (
+    matches,
+    register_quant_pattern,
+    get_quant_patterns,
+    register_dynamic_pattern,
+    get_dynamic_quant_patterns,
+)
+
+from .utils import _parent_name
+
+from abc import ABC, abstractmethod
+import copy
+import enum
+import operator
+
+# Quantization type (dynamic quantization, static quantization).
+# Should match the c++ enum in quantization_type.h
+class QuantType(enum.IntEnum):
+    DYNAMIC = 0
+    STATIC = 1
+    QAT = 2
+
+# ------------------------
+# Helper Functions
+# ------------------------
+def get_qparams(activation_post_process):
+    scale, zero_point = activation_post_process.calculate_qparams()
+    scale = float(scale)
+    zero_point = int(zero_point)
+    dtype = activation_post_process.dtype
+    return scale, zero_point, dtype
+
+def quantize_node(node, activation_post_process):
+    scale, zero_point, dtype = get_qparams(activation_post_process)
+    return torch.quantize_per_tensor(node, scale, zero_point, dtype)
+
+def quantize(quantizer, node):
+    quantize_node(node, quantizer.activation_post_process_map[node.name])
+
+# A dictionary for querying the weight index for a given op
+WEIGHT_INDEX_DICT = {
+    torch.nn.functional.conv2d : [1],
+    torch.nn.functional.linear : [1],
+}
+
+# Pattern Registrations
+
+# 1. Post Training Static Quantization and Quantization Aware Training Patterns
+
+# Note: The order of patterns is important! match function will take whatever is matched first, so we'll
+# need to put the fusion patterns before single patterns. For example, add_relu should be registered come before relu.
+# decorators are applied in the reverse order we see. Also when we match the nodes in the graph with these patterns,
+# we'll start from the last node of the graph and traverse back.
+
+# Base Pattern
+class QuantizerPattern(ABC):
+    def __init__(self, quantizer, node):
+        self.all_tensors = True
+
+    @abstractmethod
+    def convert(self, quantizer, node, load_arg, debug=False):
+        return NotImplemented
+
+@register_quant_pattern(operator.add)
+@register_quant_pattern((torch.nn.ReLU, operator.add))
+@register_quant_pattern((torch.nn.functional.relu, operator.add))
+class Add(QuantizerPattern):
```

Comment:
I think for all of these an example (input X, output Y) would be helpful for people to understand the API without having to read every single line of code

Reply:
Sure

Reply:
will probably do this in the end after everything is done, since the code might be changing.

Reply:
I think it's important to do this sooner than later, because the code is harder to read than it needs to be without the docs.  Adding documentation now and adjusting it when things change should be a net benefit, because engineers will spend time reading this code.

Reply:
Sure